cranelift-codegen/shared/src/isa/x86/encoding_bits.rs

```diff
@@ -0,0 +1,193 @@
+    Mp2_f2_0f = 0b0111,
+    Op3_0f_38 = 0b1000,
+    Mp3_66_0f_38 = 0b1001,
+    Mp3_f3_0f_38 = 0b1010,
+    Mp3_f2_0f_38 = 0b1011,
+    Op3_0f_3a = 0b1100,
+    Mp3_66_0f_3a = 0b1101,
+    Mp3_f3_0f_3a = 0b1110,
+    Mp3_f2_0f_3a = 0b1111,
+}
+
+impl From<u8> for OpcodePrefix {
+    fn from(n: u8) -> OpcodePrefix {
+        use OpcodePrefix::*;
+        match n {
+            0b0000 => Op1,
+            0b0001 => Mp1_66,
+            0b0010 => Mp1_f3,
+            0b0011 => Mp1_f2,
+            0b0100 => Op2_0f,
+            0b0101 => Mp2_66_0f,
+            0b0110 => Mp2_f3_0f,
+            0b0111 => Mp2_f2_0f,
+            0b1000 => Op3_0f_38,
+            0b1001 => Mp3_66_0f_38,
+            0b1010 => Mp3_f3_0f_38,
+            0b1011 => Mp3_f2_0f_38,
+            0b1100 => Op3_0f_3a,
+            0b1101 => Mp3_66_0f_3a,
+            0b1110 => Mp3_f3_0f_3a,
+            0b1111 => Mp3_f2_0f_3a,
+            _ => {
+                panic!("unexpected opcode prefix encoding: {:?}", n);
+            }
+        }
+    }
+}
+
+impl OpcodePrefix {
+    /// Extracts the OpcodePrefix from the opcode.
+    pub fn from_opcode(op_bytes: &[u8]) -> OpcodePrefix {
+        assert!(!op_bytes.is_empty(), "at least one opcode byte");
+
+        let prefix_bytes = &op_bytes[..op_bytes.len() - 1];
+        match prefix_bytes {
+            [] => OpcodePrefix::Op1,
+            [0x66] => OpcodePrefix::Mp1_66,
+            [0xf3] => OpcodePrefix::Mp1_f3,
+            [0xf2] => OpcodePrefix::Mp1_f2,
+            [0x0f] => OpcodePrefix::Op2_0f,
+            [0x66, 0x0f] => OpcodePrefix::Mp2_66_0f,
+            [0xf3, 0x0f] => OpcodePrefix::Mp2_f3_0f,
+            [0xf2, 0x0f] => OpcodePrefix::Mp2_f2_0f,
+            [0x0f, 0x38] => OpcodePrefix::Op3_0f_38,
+            [0x66, 0x0f, 0x38] => OpcodePrefix::Mp3_66_0f_38,
+            [0xf3, 0x0f, 0x38] => OpcodePrefix::Mp3_f3_0f_38,
+            [0xf2, 0x0f, 0x38] => OpcodePrefix::Mp3_f2_0f_38,
+            [0x0f, 0x3a] => OpcodePrefix::Op3_0f_3a,
+            [0x66, 0x0f, 0x3a] => OpcodePrefix::Mp3_66_0f_3a,
+            [0xf3, 0x0f, 0x3a] => OpcodePrefix::Mp3_f3_0f_3a,
+            [0xf2, 0x0f, 0x3a] => OpcodePrefix::Mp3_f2_0f_3a,
+            _ => {
+                panic!("unexpected opcode sequence: {:?}", op_bytes);
+            }
+        }
+    }
+
+    /// Returns the recipe name prefix.
+    ///
+    /// At the moment, each similar OpcodePrefix group is given its own Recipe.
+    /// In order to distinguish them, this string is prefixed.
+    pub fn recipe_name_prefix(self) -> &'static str {
+        use OpcodePrefix::*;
+        match self {
+            Op1 => "Op1",
+            Op2_0f => "Op2",
+            Op3_0f_38 | Op3_0f_3a => "Op3",
+            Mp1_66 | Mp1_f3 | Mp1_f2 => "Mp1",
+            Mp2_66_0f | Mp2_f3_0f | Mp2_f2_0f => "Mp2",
+            Mp3_66_0f_38 | Mp3_f3_0f_38 | Mp3_f2_0f_38 => "Mp3",
+            Mp3_66_0f_3a | Mp3_f3_0f_3a | Mp3_f2_0f_3a => "Mp3",
+        }
+    }
+}
```

Comment:
I think the approach of factoring out `EncodingBits` and `OpcodePrefix` is great for code clarity! I wish there were some way to avoid the multiple back-and-forth conversions for `OpcodePrefix` and the manual bit-fiddling in `EncodingBits`--I mean, it's better that it's all in one place but I can't be sure that I didn't miss a typo as I look through it. What do you think of https://crates.io/crates/bitflags or https://crates.io/crates/bitfields? Or could we write some tests to make it more clear that we have covered all the cases?

Reply:
Yeah, I looked at some external crates built for this purpose, but wasn't sure if we wanted to add an external dependency just for this, since that code is likely to be very static.

Reply:
Maybe a couple of doc tests then?